source/MaterialXGenShader/Nodes/SourceCodeNode.cpp

```diff
@@ -105,13 +101,41 @@ void SourceCodeNode::emitFunctionCall(const ShaderNode& node, GenContext& contex
                 throw ExceptionShaderGenError("Could not find an input named '" + variable +
                     "' on node '" + node.getName() + "'");
             }
-            shadergen.emitInput(context, input, shader);
+
+            if (input->connection)
+            {
+                string inputStr;
+                shadergen.getInput(context, input, inputStr);
+                code.push_back(inputStr);
+            }
+            else
+            {
+                string variableName = node.getName() + "_" + input->name + "_tmp";
+                Shader::VariablePtr newVariable = Shader::Variable::create(input->type, variableName, EMPTY_STRING, EMPTY_STRING, input->value);
```

Comment:
You don't need to create a new variable until you pass the if() check at 116. I think if you have duplicates then it will create these temporaries more than once. Also I don't even think you need to create a shared pointer if the Variable has a public constructor.